### PR TITLE
LED commands

### PR DESCRIPTION
**Pull request by jackpot51**
Commands will be used by keyboard-configurator in https://github.com/pop-os/keyboard-configurator/pull/22

### CODE REVIEW

**Comment by ids1024** [comment]
With this API, a client can determine if there's a controllable keyboard backlight by checking if the max brightness is non-zero, right?

Is there anyway to determine whether or not RGB color is supported, since some models may have a monochrome backlight?

**Comment by jackpot51** [comment]
@ids1024 I could adjust the led_get_color API to return a bool for whether the color can be changed, if that works

**Comment by jackpot51** [comment]
I'm running quite low on EC space on the Launch keyboard so I think it would be best if the keyboard definition said wether it supported color

**Comment by ids1024** [comment]
> so I think it would be best if the keyboard definition said wether it supported color

I suppose that works too, since it's already assumed the configurator includes specifications for different keyboard models.

File: src/board/system76/oryp5/kbled.c [hunk 1]
```diff
@@ -1,64 +0,0 @@
...
-
-void kbled_set(uint8_t level) {
-    kbled_i2c_set(0x12, &level, 1);
-}
-
-void kbled_set_color(uint32_t color) {
-    // Set blue component
-    uint8_t value = (uint8_t)(color);
-    kbled_i2c_set(0x02, &value, 1);
-    kbled_i2c_set(0x05, &value, 1);
```
Comment (crawfxrd) [inline src/board/system76/oryp5/kbled.c:50]: The addresses for oryp5 are different from darp5.

File: src/board/system76/common/kbled/rgb_pwm.c [hunk 2]
```diff
@@ -19,11 +19,28 @@ uint8_t kbled_get(void) {
...
 }
 
+uint32_t kbled_get_color(void) {
+    // Get PWM of blue component
+    uint32_t color = (uint32_t)DCR7;
+
+    // Get PWM of green component
+    color |= ((uint32_t)DCR6) << 8;
+
+    // Get PWM of green component
```
Comment (crawfxrd) [inline src/board/system76/common/kbled/rgb_pwm.c:38]: red component

File: src/board/system76/common/kbled/rgb_i2c.c [hunk 3]
```diff
@@ -39,10 +39,31 @@ uint8_t kbled_get(void) {
...
 }
 
+uint32_t kbled_get_color(void) {
+    // Get blue component from channel 0
+    uint8_t value;
+    kbled_i2c_get(0x02, &value, 1);
+    uint32_t color = (uint32_t)value;
+
+    // Get green component from channel 3
+    kbled_i2c_get(0x05, &value, 1);
```
Comment (crawfxrd) [inline src/board/system76/common/kbled/rgb_i2c.c:57]: Same issue as set_color for darp5/oryp5: addresses for color components are different.

File: src/board/system76/oryp5/kbled.c (hunk 1 above)
Comment (jackpot51) [reply src/board/system76/oryp5/kbled.c:50]: Thanks for finding this, I'll try and adjust

Comment (jackpot51) [reply src/board/system76/oryp5/kbled.c:50]: I split the two into a darp5 and an oryp5 kbled file. Hopefully Clevo has given up on this strange design so we mostly use rgb_pwm in the future

**Comment by ids1024** [comment]
Another useful feature here, that I had working through https://github.com/pop-os/system76-power/pull/190 (using inotify), is notification of changes. So the Configurator color and brightness could show changes made through keybindings or other methods.

Though that may complicate the API here too much.

**Comment by jackpot51** [comment]
@ids1024 we can't really do that without a kernel driver to handle interrupts. I recommend that the configurator polls at a reasonable interval, maybe 1 second?

**Pull request merged**